In [1]:
import numpy as np
import pandas as pd
import scipy as sp

read CSV file

In [2]:
df = pd.read_csv('Interview.csv')

set pandas option to display all column (instead of just few of them to fit the page)
If "None" is replaced by a number, that will be the number of columns will appears on the screen

In [3]:
pd.set_option('display.max_columns', None)

In [4]:
# data['new_profit_3'] = data.Profit + data.Population

display the first/head 3 rows of the table

In [5]:
df.head(3)

Date of Interview Client name         Industry Location  \
0        13.02.2015     Hospira  Pharmaceuticals  Chennai   
1        13.02.2015     Hospira  Pharmaceuticals  Chennai   
2        13.02.2015     Hospira  Pharmaceuticals  Chennai   

  Position to be closed Nature of Skillset    Interview Type Name(Cand ID)  \
0   Production- Sterile            Routine  Scheduled Walkin   Candidate 1   
1   Production- Sterile            Routine  Scheduled Walkin   Candidate 2   
2   Production- Sterile            Routine  Scheduled Walkin   Candidate 3   

  Gender Candidate Current Location Candidate Job Location Interview Venue  \
0   Male                    Chennai                  Hosur           Hosur   
1   Male                    Chennai              Bangalore           Hosur   
2   Male                    Chennai                Chennai           Hosur   

  Candidate Native location  \
0                     Hosur   
1                    Trichy   
2                   Chennai   

  Have you obtained the necessary permission to start at the required time  \
0                                                Yes                         
1                                                Yes                         
2                                                NaN                         

  Hope there will be no unscheduled meetings  \
0                                        Yes   
1                                        Yes   
2                                         Na   

  Can I Call you three hours before the interview and follow up on your attendance for the interview  \
0                                                Yes                                                   
1                                                Yes                                                   
2                                                NaN                                                   

  Can I have an alternative number/ desk number. I assure you that I will not trouble you too much  \
0                                                Yes                                                 
1                                                Yes                                                 
2                                                NaN                                                 

  Have you taken a printout of your updated resume. Have you read the JD and understood the same  \
0                                                Yes                                               
1                                                Yes                                               
2                                                NaN                                               

  Are you clear with the venue details and the landmark.  \
0                                                Yes       
1                                                Yes       
2                                                NaN       

  Has the call letter been shared Expected Attendance Observed Attendance  \
0                             Yes                 Yes                  No   
1                             Yes                 Yes                  No   
2                             NaN           Uncertain                  No   

  Marital Status  Unnamed: 23  Unnamed: 24  Unnamed: 25  Unnamed: 26  \
0         Single          NaN          NaN          NaN          NaN   
1         Single          NaN          NaN          NaN          NaN   
2         Single          NaN          NaN          NaN          NaN   

   Unnamed: 27  
0          NaN  
1          NaN  
2          NaN

## First analyzing of the data
data.shape = number of (rows, columns) of the data
data.describe() = basic assumptions about the data

In [6]:
df.shape

(1234, 28)

if we use describe on the raw data, many NaN (Not a Number) data is display, those can be deleted.

In [7]:
df.describe()

Unnamed: 23  Unnamed: 24  Unnamed: 25  Unnamed: 26  Unnamed: 27
count          0.0          0.0          0.0          0.0          0.0
mean           NaN          NaN          NaN          NaN          NaN
std            NaN          NaN          NaN          NaN          NaN
min            NaN          NaN          NaN          NaN          NaN
25%            NaN          NaN          NaN          NaN          NaN
50%            NaN          NaN          NaN          NaN          NaN
75%            NaN          NaN          NaN          NaN          NaN
max            NaN          NaN          NaN          NaN          NaN

## Delete excess rows and columns
in order to delete, we use a pandas function: drop.

In [8]:
df.drop(['Unnamed: 23','Unnamed: 24','Unnamed: 25','Unnamed: 26','Unnamed: 27'],axis=1)
df.head(1)

Date of Interview Client name         Industry Location  \
0        13.02.2015     Hospira  Pharmaceuticals  Chennai   

  Position to be closed Nature of Skillset    Interview Type Name(Cand ID)  \
0   Production- Sterile            Routine  Scheduled Walkin   Candidate 1   

  Gender Candidate Current Location Candidate Job Location Interview Venue  \
0   Male                    Chennai                  Hosur           Hosur   

  Candidate Native location  \
0                     Hosur   

  Have you obtained the necessary permission to start at the required time  \
0                                                Yes                         

  Hope there will be no unscheduled meetings  \
0                                        Yes   

  Can I Call you three hours before the interview and follow up on your attendance for the interview  \
0                                                Yes                                                   

  Can I have an alternative number/ desk number. I assure you that I will not trouble you too much  \
0                                                Yes                                                 

  Have you taken a printout of your updated resume. Have you read the JD and understood the same  \
0                                                Yes                                               

  Are you clear with the venue details and the landmark.  \
0                                                Yes       

  Has the call letter been shared Expected Attendance Observed Attendance  \
0                             Yes                 Yes                  No   

  Marital Status  Unnamed: 23  Unnamed: 24  Unnamed: 25  Unnamed: 26  \
0         Single          NaN          NaN          NaN          NaN   

   Unnamed: 27  
0          NaN

Caution: those changes that made by drop function will not effect the data because it works on the "pass by value" data. Attribute 'inplace=True' will solve this)

In [9]:
df.drop(['Unnamed: 23','Unnamed: 24','Unnamed: 25','Unnamed: 26','Unnamed: 27'],axis=1,inplace=True)
df.head(1)

Date of Interview Client name         Industry Location  \
0        13.02.2015     Hospira  Pharmaceuticals  Chennai   

  Position to be closed Nature of Skillset    Interview Type Name(Cand ID)  \
0   Production- Sterile            Routine  Scheduled Walkin   Candidate 1   

  Gender Candidate Current Location Candidate Job Location Interview Venue  \
0   Male                    Chennai                  Hosur           Hosur   

  Candidate Native location  \
0                     Hosur   

  Have you obtained the necessary permission to start at the required time  \
0                                                Yes                         

  Hope there will be no unscheduled meetings  \
0                                        Yes   

  Can I Call you three hours before the interview and follow up on your attendance for the interview  \
0                                                Yes                                                   

  Can I have an alternative number/ desk number. I assure you that I will not trouble you too much  \
0                                                Yes                                                 

  Have you taken a printout of your updated resume. Have you read the JD and understood the same  \
0                                                Yes                                               

  Are you clear with the venue details and the landmark.  \
0                                                Yes       

  Has the call letter been shared Expected Attendance Observed Attendance  \
0                             Yes                 Yes                  No   

  Marital Status  
0         Single

In [10]:
df.shape

(1234, 23)

In [11]:
#df.columns

After deletion, new dataframe will have 23 columns left, now we assign new names to these all 23 columns by create a list of all the names then assign it to data.columns

In [12]:
columns = ['date','name','industry','job_location','position', 'skill','type','candidate_id','gender',
           'candidate_address','candidate_job_location','venue','nation','is_ready','in_scheduled_meetings',
           'prev_call','additional_number','understand_jd','know_location','letter_received',
           'expected_attendance','observed_attendance','marital_status']
df.columns = columns
df.head(1)

date     name         industry job_location             position  \
0  13.02.2015  Hospira  Pharmaceuticals      Chennai  Production- Sterile   

     skill              type candidate_id gender candidate_address  \
0  Routine  Scheduled Walkin  Candidate 1   Male           Chennai   

  candidate_job_location  venue nation is_ready in_scheduled_meetings  \
0                  Hosur  Hosur  Hosur      Yes                   Yes   

  prev_call additional_number understand_jd know_location letter_received  \
0       Yes               Yes           Yes           Yes             Yes   

  expected_attendance observed_attendance marital_status  
0                 Yes                  No         Single

df.describe(include=['O'])). To select pandas categorical columns, use 'category'. Which is a must-be-used-mode for non-number values. 'O' stands for "objects"

In [13]:
df.describe(include=['O'])

date                     name industry job_location position  \
count         1233                     1234     1233         1233     1233   
unique          96                       16        7           11        7   
top     06.02.2016  Standard Chartered Bank     BFSI      Chennai  Routine   
freq           220                      904      949          754     1023   

                             skill               type   candidate_id gender  \
count                         1233               1233           1233   1233   
unique                          92                  6           1233      2   
top     JAVA/J2EE/Struts/Hibernate  Scheduled Walk In  Candidate 257   Male   
freq                           220                456              1    965   

       candidate_address candidate_job_location    venue   nation is_ready  \
count               1233                   1233     1233     1233     1029   
unique                10                      7        7       46        7   
top              Chennai                Chennai  Chennai  Chennai      Yes   
freq                 754                    893      852      595      917   

       in_scheduled_meetings prev_call additional_number understand_jd  \
count                    986       986               986           985   
unique                     7         5                 6             8   
top                      Yes       Yes               Yes           Yes   
freq                     949       951               936           940   

       know_location letter_received expected_attendance observed_attendance  \
count            985             988                1228                1233   
unique             7              12                   7                   8   
top              Yes             Yes                 Yes                 Yes   
freq             946             932                 882                 701   

       marital_status  
count            1233  
unique              2  
top            Single  
freq              767

Analyze: as in df.shape, we have 1234 rows, but only 'name' column that have 1234 filled item. Other columns are all missing at least one item. Which might leads to a guess that there is a row that have no value but name cell. In practise, those kind of row will be put at the end of the table. Then we will check the last rows

In [14]:
df.tail(2)

date    name         industry job_location position  \
1232  06.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   
1233         NaN                    NaN          NaN      NaN   

                   skill        type    candidate_id  gender  \
1232  generic drugs – RA  Scheduled   Candidate 1233  Female   
1233                 NaN         NaN             NaN     NaN   

     candidate_address candidate_job_location    venue   nation is_ready  \
1232           Chennai                Chennai  Chennai  Kolkata      NaN   
1233               NaN                    NaN      NaN      NaN      NaN   

     in_scheduled_meetings prev_call additional_number understand_jd  \
1232                   NaN       NaN               NaN           NaN   
1233                   NaN       NaN               NaN           NaN   

     know_location letter_received expected_attendance observed_attendance  \
1232           NaN             NaN           Uncertain                 Yes   
1233           NaN             NaN                 NaN                 NaN   

     marital_status  
1232         Single  
1233            NaN

as can be seen, the row indexed 1233 (order 1234) has all NaN cells, the cell 'name' is empty which won't affect the outcome of the analyzing process. => this row can be deleted.

In [15]:
# select date row, then select the first 2 rows
df['name'][:2]

0    Hospira
1    Hospira
Name: name, dtype: object

In [16]:
# select 2 last rows
df[-2:]

date    name         industry job_location position  \
1232  06.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   
1233         NaN                    NaN          NaN      NaN   

                   skill        type    candidate_id  gender  \
1232  generic drugs – RA  Scheduled   Candidate 1233  Female   
1233                 NaN         NaN             NaN     NaN   

     candidate_address candidate_job_location    venue   nation is_ready  \
1232           Chennai                Chennai  Chennai  Kolkata      NaN   
1233               NaN                    NaN      NaN      NaN      NaN   

     in_scheduled_meetings prev_call additional_number understand_jd  \
1232                   NaN       NaN               NaN           NaN   
1233                   NaN       NaN               NaN           NaN   

     know_location letter_received expected_attendance observed_attendance  \
1232           NaN             NaN           Uncertain                 Yes   
1233           NaN             NaN                 NaN                 NaN   

     marital_status  
1232         Single  
1233            NaN

In [17]:
df.describe(percentiles=[0.1,0.25,0.5,0.75,0.9,0.95,0.99])

date                     name industry job_location position  \
count         1233                     1234     1233         1233     1233   
unique          96                       16        7           11        7   
top     06.02.2016  Standard Chartered Bank     BFSI      Chennai  Routine   
freq           220                      904      949          754     1023   

                             skill               type   candidate_id gender  \
count                         1233               1233           1233   1233   
unique                          92                  6           1233      2   
top     JAVA/J2EE/Struts/Hibernate  Scheduled Walk In  Candidate 257   Male   
freq                           220                456              1    965   

       candidate_address candidate_job_location    venue   nation is_ready  \
count               1233                   1233     1233     1233     1029   
unique                10                      7        7       46        7   
top              Chennai                Chennai  Chennai  Chennai      Yes   
freq                 754                    893      852      595      917   

       in_scheduled_meetings prev_call additional_number understand_jd  \
count                    986       986               986           985   
unique                     7         5                 6             8   
top                      Yes       Yes               Yes           Yes   
freq                     949       951               936           940   

       know_location letter_received expected_attendance observed_attendance  \
count            985             988                1228                1233   
unique             7              12                   7                   8   
top              Yes             Yes                 Yes                 Yes   
freq             946             932                 882                 701   

       marital_status  
count            1233  
unique              2  
top            Single  
freq              767

In [18]:
# df.iloc(1233) is not accessible, raise exception
df1 = df.iloc[1233]

In [19]:
# delete the data then display the last 2 rows
# to check the output
df.drop(df.index[1233], inplace=True)
df.tail(2)

date    name         industry job_location position  \
1231  06.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   
1232  06.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   

                   skill        type    candidate_id  gender  \
1231  generic drugs – RA  Scheduled   Candidate 1232    Male   
1232  generic drugs – RA  Scheduled   Candidate 1233  Female   

     candidate_address candidate_job_location    venue   nation is_ready  \
1231           Chennai                Chennai  Chennai  Chennai      Yes   
1232           Chennai                Chennai  Chennai  Kolkata      NaN   

     in_scheduled_meetings prev_call additional_number understand_jd  \
1231                   Yes       Yes               Yes           Yes   
1232                   NaN       NaN               NaN           NaN   

     know_location letter_received expected_attendance observed_attendance  \
1231           Yes             Yes                 Yes                 Yes   
1232           NaN             NaN           Uncertain                 Yes   

     marital_status  
1231         Single  
1232         Single

In [20]:
# recheck the result of describe method after deletion
df.describe(include=['O'])

date                     name industry job_location position  \
count         1233                     1233     1233         1233     1233   
unique          96                       15        7           11        7   
top     06.02.2016  Standard Chartered Bank     BFSI      Chennai  Routine   
freq           220                      904      949          754     1023   

                             skill               type   candidate_id gender  \
count                         1233               1233           1233   1233   
unique                          92                  6           1233      2   
top     JAVA/J2EE/Struts/Hibernate  Scheduled Walk In  Candidate 257   Male   
freq                           220                456              1    965   

       candidate_address candidate_job_location    venue   nation is_ready  \
count               1233                   1233     1233     1233     1029   
unique                10                      7        7       46        7   
top              Chennai                Chennai  Chennai  Chennai      Yes   
freq                 754                    893      852      595      917   

       in_scheduled_meetings prev_call additional_number understand_jd  \
count                    986       986               986           985   
unique                     7         5                 6             8   
top                      Yes       Yes               Yes           Yes   
freq                     949       951               936           940   

       know_location letter_received expected_attendance observed_attendance  \
count            985             988                1228                1233   
unique             7              12                   7                   8   
top              Yes             Yes                 Yes                 Yes   
freq             946             932                 882                 701   

       marital_status  
count            1233  
unique              2  
top            Single  
freq              767

## Standardize data
df.column_name.value_counts() show the frequency of the data in categories in the table

#### date

In [21]:
df.date.value_counts()

06.02.2016               220
04/12/16                  86
13.04.2016                84
02.09.2016                43
25.06.2015                43
21.4.16                   38
22.01.2016                34
07.05.2016                31
03.04.2015                30
14.01.2016                28
26.02.2016                26
27.02.2016                26
20.08.2016                25
05/11/2016                25
25.5.2016                 24
23.02.2016                22
02.12.2015                19
19.06.2015                18
19.03.16                  17
30.01.2016                16
09-01-2016                15
13.06.2016                15
11.5.16                   14
21.4.2016                 14
25-5-2016                 14
30.1.16                   14
22.03.2016                13
13/03/2015                13
3.6.16                    12
25.05.16                  12
                        ... 
28.08.2016 & 9.30 AM       2
28.08.2016 & 09.30 AM      2
28.08.2016 & 11.30 AM      2
19 Apr 16     

In [22]:
def convert_dmy(date):
    s1 = date.replace('-',' ').replace('/',' ').replace('.',' ')
    
    # replace all duplicated space and turn into one space
    s2 = " ".join(s1.split())
    d, m, y = s2.split(" ")
    if len(d) == 1:
        d = "0"+d
    else:
        pass

    if len(m) == 1:
        m = "0"+m
    else:
        pass

    if len(y) == 2:
        y = "20" + y
    else:
        pass
    return d,m,y

def standardize_date(df):
    for cell in df['date']:
        # & in cell means the string contains time
        if "&" in cell:
            date, time = cell.split('&')
            d, m, y = convert_dmy(date)
            cell = d + "/" + m + "/" + y
            
        else:
            d, m, y = convert_dmy(cell)
            cell = d + "/" + m + "/" + y

In [23]:
standardize_date(df)

ValueError: too many values to unpack (expected 3)

#### job_location

In [ ]:
df.job_location.value_counts()

In [ ]:
# lower case the all values
df['job_location'] = df['job_location'].str.lower()
# repalce "chennai " to "chennai"
df.loc[df['job_location'] == 'chennai ', 'job_location'] = 'chennai'
# repalce "gurgaonr" to "gurgaon"
df.loc[df['job_location'] == 'gurgaonr', 'job_location'] = 'gurgaon'
# repalce "- cochin-" to "cochin"
df.loc[df['job_location'] == '- cochin- ', 'job_location'] = 'cochin'

In [ ]:
df['job_location'].value_counts()